agents/builtin/mysql/mysql.go

```diff
@@ -93,78 +114,179 @@ func (m *MySQL) Run(ctx context.Context) {
 			return
 
 		case <-t.C:
-			request, err := m.get(db.Querier)
+			if !running {
+				m.changes <- Change{Status: inventorypb.AgentStatus_STARTING}
+			}
+
+			buckets, err := m.getBuckets(start, wait)
+
+			start = time.Now().Truncate(0) // strip monotoning clock reading
+			wait = start.Truncate(querySummaries).Add(querySummaries).Sub(start)
+			m.l.Debugf("Scheduling next collection in %s at %s.", wait, start.Add(wait))
+			t.Reset(wait)
+
 			if err != nil {
 				m.l.Error(err)
 				running = false
 				m.changes <- Change{Status: inventorypb.AgentStatus_WAITING}
-				time.Sleep(time.Second)
-				m.changes <- Change{Status: inventorypb.AgentStatus_STARTING}
 				continue
 			}
 
 			if !running {
-				m.changes <- Change{Status: inventorypb.AgentStatus_RUNNING}
 				running = true
+				m.changes <- Change{Status: inventorypb.AgentStatus_RUNNING}
 			}
 
-			select {
-			case <-ctx.Done():
-				t.Stop()
-				break
-			case m.changes <- Change{Request: request}:
-				// nothing
-			}
+			m.changes <- Change{Request: &qanpb.CollectRequest{MetricsBucket: buckets}}
 		}
 	}
 }
 
-func (m *MySQL) get(q *reform.Querier) (qanpb.CollectRequest, error) {
-	var res qanpb.CollectRequest
-	structs, err := q.SelectAllFrom(eventsStatementsSummaryByDigestView, "")
+func (m *MySQL) runHistoryCacheRefresher(ctx context.Context) {
+	t := time.NewTicker(refreshHistory)
+	defer t.Stop()
+
+	for {
+		if err := m.refreshHistoryCache(); err != nil {
+			m.l.Error(err)
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			// nothing, continue loop
+		}
+	}
+}
+
+func (m *MySQL) refreshHistoryCache() error {
+	current, err := getHistory(m.db.Querier)
 	if err != nil {
-		return res, err
+		return err
+	}
+	m.historyCache.refresh(current)
+	return nil
+}
+
+func (m *MySQL) getBuckets(periodStart time.Time, periodLength time.Duration) ([]*qanpb.MetricsBucket, error) {
+	current, err := getSummaries(m.db.Querier)
+	if err != nil {
+		return nil, err
+	}
+
+	buckets := makeBuckets(current, m.summaryCache.get(), m.l)
+
+	m.summaryCache.refresh(current)
+
+	// add examples from history cache
+	history := m.historyCache.get()
+	for i, b := range buckets {
+		if esh := history[b.Queryid]; esh != nil {
+			// TODO test if we really need that
+			if b.DSchema == "" {
+				b.DSchema = pointer.GetString(esh.CurrentSchema)
+			}
+
+			if esh.SQLText != nil {
+				b.Example = *esh.SQLText
+				b.ExampleFormat = qanpb.ExampleFormat_EXAMPLE
+				b.ExampleType = qanpb.ExampleType_RANDOM
+			}
+		}
+
+		buckets[i] = b
 	}
 
-	for _, str := range structs {
-		ess := str.(*eventsStatementsSummaryByDigest)
+	startS := uint32(periodStart.Unix())
+	lengthS := uint32(periodLength.Seconds())
+	for i, b := range buckets {
```

Comment:
why do we use two loops instead of one?

Reply:
Good point, that's a missed refactoring opportunity after functions were merged.